#93: gprproject/__init__.py, gprproject/gprbuild.py: python builder: implementation of --gpr-opts

dsauvage opened the pull request:
See #93

Jicquel commented on gprproject/gprbuild.py:
```diff
@@ -194,6 +201,10 @@ def run(self, args: list[str], **kwargs) -> int:
         for key, value in self.variables.items():
             cmd.append(f"-X{key}={value}")
 
+        if self.gpr_opts is not None:
```
This block of code can be moved into the `if cmd_name == "gprbuild":` block above. Otherwise, arguments may be incorrectly added to gprinstall, which supports fewer arguments than gprbuild. Based on the parameter description, I assumed that `gpr_opts` is intended to be used exclusively for gprbuild.

dsauvage replied:
The `gpr_opts` option was already in the code base but not implemented. It is not specific to gprbuild, but should be available to gprbuild, gprclean and gprinstall.

Jicquel commented on gprproject/gprbuild.py:
```diff
@@ -277,6 +288,7 @@ def save(self):
             "symcc": self.symcc,
             "prefix": self.prefix,
             "gpr_paths": self.gpr_paths,
+            "gpr_opts": self.gpr_opts,
```
In response to https://github.com/AdaCore/gnatcoll-core/pull/94/files#r1881804015:
To make it work properly, this line needs to be removed. Each time the build command is run, its environment and provided arguments are saved in a JSON file for reuse by the clean, uninstall, and install commands. You can search for "load" in `gprproject/init.py` for reference. If we store `gpr_opts` arguments here, they may not be valid for the other tools. For instance:

```
$ gnatcoll_core.gpr.py build --gpr-opts="-O0"
$ gnatcoll_core.gpr.py install <<---- raises gprinstall: unrecognized option '-O0'
```

Jicquel replied:
This is a good solution! I'll review your new PR if you plan to create one.

Jicquel commented on gprproject/gprbuild.py:
```diff
@@ -70,6 +71,7 @@ def __init__(
         :param jobs: level of parallelism for gpr tools that support it
         :param gnatcov: if True add gnatcov instrumentation
         :param symcc: if True add symcc instrumentation
+        :param gpr_opts: arguments passed to gprbuild
```
Could you update to something like: "arguments passed to gpr tools"

dsauvage commented:
In our case, as we needed arguments to be passed to all the gpr tools (dealing with multiple Ada runtime context), we implemented this feature using the existing `--gpr-opts` argument that was not implemented. 

Since then we implemented a `--rts` argument to manage builds with multiple Ada runtime context. The generated project json file stores multiple json project configuration object (one per Ada runtime). I can plan to publish the corresponding branch if this is of interest for you.

Seems a bug has been fixed in GNAT-FSF-14, we might not need to use the `--gpr-opts` above anymore.

Jicquel commented:
Hi, if it does not take you time, we would indeed be happy to take a look at your --rts implementation.

dsauvage commented:
> Hi, if it does not take you time, we would indeed be happy to take a look at your --rts implementation.

Sure,
kindly go through [Commit f2dc363](https://github.com/AdaLabs/gnatcoll-core/commit/f2dc363ec16013c790e1b5f6c0e262bc6ef347ca) on our forked gnatcoll-core branch [25.1-adalabs](https://github.com/AdaLabs/gnatcoll-core/compare/25.1-adalabs)

Jicquel commented:
Thank you, we will give it a look and come back to you
